### ISSUE TITLE
DOOM 64 sound crackling

### ISSUE THREAD

**Issue by Tasosgemah** [libretro/mupen64plus-libretro-nx#72]
When using this core, DOOM 64 has sound issues when you are in-game (crackling/popping).

In the other cores (older Mupen and ParaLLel) the sound is clean and the same applies with the standalone.

Tried disabling anything that could slow down emulation (like GPU sync) but the crackling doesn't stop.

**Comment by m4xw** [comment]
In RA go "Help" -> "Send Debug Info", give me your ID.

**Comment by Tasosgemah** [comment]
Well, i screwed up. I got a number but because my memory sucks i wanted to write it down but the message went away. Now when i try to take the ID it gives me an error message that it failed to send to the server.

**Comment by Tasosgemah** [comment]
14252

**Comment by m4xw** [comment]
Try Audio latency 85ms or turn threaded video on.

**Comment by Tasosgemah** [comment]
I tried them both independently and together. I increased audio latency to 128 as well.

It's still popping for me. Less, but it's still there.

**Comment by m4xw** [comment]
Can you check this against mupen+gliden updated standalone?

**Comment by Tasosgemah** [comment]
Yes. Doom64 sound is fine using Mupen standalone and GlideN64 release 4.

**Comment by m4xw** [comment]
Thanks

**Comment by twinaphex** [comment]
I can reproduce the same issue. This definitely does not happen with Parallel N64.

**Comment by mudlord** [comment]
Works fine here with my own libretro loader.

**Comment by m4xw** [comment]
@mudlord can u check your audio buffer saturation?

**Comment by mudlord** [comment]
| Name | Value | Type
-- | -- | -- | --
  | outsamples_max | 196608 | unsigned __int64


  | Name | Value | Type
-- | -- | -- | --
  | sampsize | 30720 | unsigned __int64
	_fifo = fifo_new(sampsize); //number of bytes

I leave plenty of samples in the tanks.

mostly get around 4k samples being written at a time.

**Comment by m4xw** [comment]
@Tasosgemah might be tied to your monitor again?

**Comment by Tasosgemah** [comment]
On my 60hz TV the sound crackles no matter what.

On my 240hz monitor i can fix it if i disable the "sync with exact frame rate option". But that option doesn't fix the sound on the 60hz TV.

**Comment by m4xw** [comment]
Seems the game syncs to 62vi/s.
Tatsuya reported simliar results on his 120hz monitor, precisely only if u disable said options....
ffs

**Comment by m4xw** [comment]
also the sample rate is odd, as well as some odd push samples.
couple that with ra's audio sync.. fun times.
standalone doesnt resample and just sets output to a close enough rate.

**Comment by m4xw** [comment]
@Tasosgemah What does actually happen if u disable vsync and have audio sync & only have vsync and disable audio sync, on 60hz.

**Comment by Tasosgemah** [comment]
vsync OFF / audio sync ON seems to fix it... with the added screen tearing.

Vsync ON / audio sync OFF still crackles

**Comment by m4xw** [comment]
oh boy

**Comment by m4xw** [comment]
@Tasosgemah time to revisit yoshi with the same test?

**Comment by m4xw** [comment]
@Tasosgemah this build might fix it https://cdn.discordapp.com/attachments/555706037412560909/675457193226534953/mupen64plus_next_libretro.7z
need to think of a solution.
now reports fps and not vi/s too, so dont wonder (also its gln)

**Comment by m4xw** [comment]
Seems to be fixed for quake64 too, try yoshi with it.

**Comment by Tasosgemah** [comment]
The sound is fixed in DOOM 64 but now the frame rate is messed up. I get constant microstuttering that wasn't there before. Even in exclusive full screen or with sync to exact frame rate enabled/disabled.

**Comment by m4xw** [comment]
Framerate isn't messed up.
Before it was reporting vi/s, now its real fps.
The Issue is doom64 is 15 samples short of maintaining a full buffer with a 50/60 vi/s callback to the frontend.

**Comment by m4xw** [comment]
Welcome to no frame duping :shrek:

**Comment by Tasosgemah** [comment]
I tested Mario 64 and now i'm getting horrible frame pacing issues at steady intervals.

I really hope this fix isn't going to be used...

**Comment by Tasosgemah** [comment]
This fix reminds me of the good old "fixed audio timing" in PJ64, in that it fixes the audio but makes most games have extra stuttering and frame pacing issues that don't exist in the real console.

**Comment by m4xw** [comment]
Welcome to real N64 pacing!

**Comment by Tasosgemah** [comment]
No, that's not how the real N64 runs. I have a real N64 right now and Mario does not have these stutters.

**Comment by m4xw** [comment]
Try your n64 on that monitor :P

**Comment by Tasosgemah** [comment]
Im using the 60hz monitor now, not the 240hz. This is not how the N64 runs, Mario 64 is smooth 30fps outside the castle.

**Comment by Tasosgemah** [comment]
On the core you uploaded i'm getting VERY bad frame pacing issues every few seconds.

**Comment by m4xw** [comment]
whats vsync set to?

**Comment by Tasosgemah** [comment]
Vsync is ON. I always have it ON because that fixes the screen tearing.

**Comment by m4xw** [comment]
Same with off?

**Comment by Tasosgemah** [comment]
I mean, just about every other game didn't have sound issues, it was only DOOM. I don't think it's worth to mangle the frame rate of every N64 game just to fix the audio in that game :P

**Comment by m4xw** [comment]
? It would only be enabled for that game and also a core option, what question is that lol

**Comment by m4xw** [comment]
Try this
[mupen64plus_next_libretro.zip](https://github.com/libretro/mupen64plus-libretro-nx/files/4173210/mupen64plus_next_libretro.zip)

**Comment by Tasosgemah** [comment]
give me a sec because i'm trying a lot of things at once :D

**Comment by Tasosgemah** [comment]
Ok now it's the same behavior as before. Frame rates are fine and DOOM's sound crackle on the 60hz monitor.

Not using the 240hz one right now at all anyway.

**Comment by m4xw** [comment]
So bsOnColorImageChange for doom it is.

**Comment by Tasosgemah** [comment]
Probably but it does make it run with an uneven frame rate. Now, i don't have DOOM right now *the real cart) but i do remember it running smooth. It also runs smooth in Parallel with no sound crackle.

**Comment by m4xw** [comment]
> 
> 
> Probably but it does make it run with an uneven frame rate. Now, i don't have DOOM right now (the real cart) but i do remember it running smooth. It also runs smooth in Parallel with no sound crackle.

its 62 vi/s with that game, so yes it does have odd vertical interrupts.

**Comment by m4xw** [comment]
There might also be a other option, we will see after the next mupen core rebase, but for right now, this will do.

**Comment by Tasosgemah** [comment]
Despite that, it runs smooth in ParaLLel+Angrylion and the real N64. I don't know why it looks so much worse on Mupen64plus-Next with the audio fix but knowing myself i wouldn't stand playing it like this. From the intro you can see as the camera moves, it makes all these little jumps. It's not right.

**Comment by m4xw** [comment]
Its because no frame duping so it doesnt sync up to your monitors refresh rate.

**Comment by Tasosgemah** [comment]
Question is how come it syncs on a real N64 or ParaLLel?

Anyway, i hope something else comes up, until then ParaLLel seems the best option for this game.

**Comment by m4xw** [comment]
Because they have a intermediate call that dupes the image by the frontend.

**Comment by m4xw** [comment]
@Tasosgemah Probably resolved with the recent mupen updates that I just pulled in:
https://git.m4xw.net/libretro/cores/mupen64plus-libretro-nx/pipelines/4974/builds

**Comment by Tasosgemah** [comment]
Tested it now and it looks great. No sound popping and no frame rate regressions so far.

**Comment by m4xw** [comment]
@Tasosgemah Worth giving yoshi a shot?

**Comment by m4xw** [comment]
Also for reference I will leave this issue open since I still want to make the swap trigger a core option.
But nice!

**Comment by m4xw** [comment]
Also I forgot to mention, this bumps savestate version so new states fail to load on current dev branch

**Comment by Tasosgemah** [comment]
Yoshi's Story has all kinds of slowdowns/stutters/frame pacing problems and all the sound issues that come with that if i use "stripped". If i use "OnePiece" it runs smoothly with no sound issues. I only tested for a few minutes on the first level though. Dunno if there are other visual bugs for going with onepiece, knowing it's the least accurate option.

**Comment by m4xw** [comment]
Oh I didn't meant to try stripped, probably just too intense for that game.
But didn't it have a issue with your 240hz monitor?

**Comment by Tasosgemah** [comment]
Runs fine with 240hz yeah.

**Comment by m4xw** [comment]
So that fixed itself with exact sync and the likes too?

**Comment by Tasosgemah** [comment]
Looks like that, yeah. I have exact sync ON as default.

**Comment by m4xw** [comment]
Sweet, thanks.
I hope i get these changes out within 48h as 2.0.5

**Comment by m4xw** [comment]
Also a sidenote, pokemon snap should work in that version with interpreter too

**Comment by Tasosgemah** [comment]
Thanks. Does Pokemon Snap need interpreter to run? I haven't tested that game much.

**Comment by m4xw** [comment]
No, but the fix that makes the red dot appear only works with interpreter

**Comment by m4xw** [comment]
Oof got a bit sidetracked, hope I wont forget it again, heh.

**Comment by m4xw** [comment]
Pushed..
